pkg/ccl/changefeedccl/changefeed_dist.go

```diff
@@ -215,127 +234,179 @@ func startDistChangefeed(
-		sql.DistributionTypeAlways)
-
-	var spanPartitions []sql.SpanPartition
-	if details.SinkURI == `` {
-		// Sinkless feeds get one ChangeAggregator on the gateway.
-		spanPartitions = []sql.SpanPartition{{SQLInstanceID: dsp.GatewayID(), Spans: trackedSpans}}
-	} else {
-		// All other feeds get a ChangeAggregator local on the leaseholder.
-		var err error
-		spanPartitions, err = dsp.PartitionSpans(ctx, planCtx, trackedSpans)
-		if err != nil {
-			return err
-		}
+
+	p, planCtx, err := makePlan(execCtx, jobID, details, initialHighWater, checkpoint, trackedSpans, selectClause)(ctx, dsp)
+	if err != nil {
+		return err
 	}
 
-	// Use the same checkpoint for all aggregators; each aggregator will only look at
-	// spans that are assigned to it.
-	// We could compute per-aggregator checkpoint, but that's probably an overkill.
-	aggregatorCheckpoint := execinfrapb.ChangeAggregatorSpec_Checkpoint{
-		Spans:     checkpoint.Spans,
-		Timestamp: checkpoint.Timestamp,
+	replanOracle := sql.ReplanOnChangedFraction(func() float64 { return replanChangefeedThreshold.Get(execCtx.ExecCfg().SV()) })
+	if knobs, ok := cfKnobs.(*TestingKnobs); ok && knobs != nil && knobs.ShouldReplan != nil {
+		replanOracle = knobs.ShouldReplan
 	}
 
-	var checkpointSpanGroup roachpb.SpanGroup
-	checkpointSpanGroup.Add(checkpoint.Spans...)
+	replanner, stopReplanner := sql.PhysicalPlanChangeChecker(ctx,
+		p,
+		makePlan(execCtx, jobID, details, initialHighWater, checkpoint, trackedSpans, selectClause),
+		execCtx,
+		replanOracle,
+		func() time.Duration { return replanChangefeedFrequency.Get(execCtx.ExecCfg().SV()) },
+	)
+
+	g := ctxgroup.WithContext(ctx)
+	g.GoCtx(func(ctx context.Context) error {
+		defer stopReplanner()
+
+		resultRows := makeChangefeedResultWriter(resultsCh)
+		recv := sql.MakeDistSQLReceiver(
+			ctx,
+			resultRows,
+			tree.Rows,
+			execCtx.ExecCfg().RangeDescriptorCache,
+			noTxn,
+			nil, /* clockUpdater */
+			evalCtx.Tracing,
+			execCtx.ExecCfg().ContentionRegistry,
+			nil, /* testingPushCallback */
+		)
+		defer recv.Release()
+
+		var finishedSetupFn func()
+		if details.SinkURI != `` {
+			// We abuse the job's results channel to make CREATE CHANGEFEED wait for
+			// this before returning to the user to ensure the setup went okay. Job
+			// resumption doesn't have the same hack, but at the moment ignores
+			// results and so is currently okay. Return nil instead of anything
+			// meaningful so that if we start doing anything with the results
+			// returned by resumed jobs, then it breaks instead of returning
+			// nonsense.
+			finishedSetupFn = func() { resultsCh <- tree.Datums(nil) }
+		}
+
+		// Copy the evalCtx, as dsp.Run() might change it.
+		evalCtxCopy := *evalCtx
+		// p is the physical plan, recv is the distsqlreceiver
+		dsp.Run(ctx, planCtx, noTxn, p, recv, &evalCtxCopy, finishedSetupFn)()
+		return resultRows.Err()
+	})
+
+	g.GoCtx(replanner)
+
+	err = g.Wait()
```

Comment:
Since we run 2 tasks, and we wait for both of them, I suggest we use
`ctxgroup.GoAndWait` as in 
```
  if err := ctxgroup.GoAndWait(ctx, func() {... your first function }, replan); err != nil {
     ....
   }
```

GoAndWait is less error prone than multiple g.GoCtx calls.
if you take the function you have above (that invokes dsp.Run) and instead do something like:
```
execPlan := func(ctx context.Context) error {
    .... 
   dsp.Run(....)
   return resultRows.Err()
}
```
then the code to invoke both go routines will be very short and very readable
(you could also move execPlan function and make it a standalone function if you want)

```
if err := ctxgroup.GoAndWait(ctx, execPlan, replan); err != nil {
  // Oh noo -- handle error
}
```